app/src/main/java/net/programmierecke/radiodroid2/alarm/AlarmReceiver.java

```diff
@@ -215,6 +217,83 @@ protected void onPostExecute(String result) {
         }.execute();
     }
 
+    private void setAlarmVolume(float volume) {
+        try {
+            if (BuildConfig.DEBUG) { Log.d(TAG, "Setting volume to "+volume); }
+            itsPlayerService.SetVolume(volume);
+        } catch (Exception e) {
+            Log.e(TAG, "Couldn't set volume "+e);
+        }
+    }
+
+    private void graduallyIncreaseAlarmVolume(Context context) {
+        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+        int slowWakeMillis = sharedPref.getInt("gradually_increase_volume", 0) * 10000;
+
+        if (slowWakeMillis == 0) {
+            if (BuildConfig.DEBUG) { Log.d(TAG, "Gradual alarm volume disabled"); }
+            return;
+        }
+
+        float maxVolume = 100f;
+        float minVolume = 0f;
+        float volumeRange = maxVolume - minVolume;
+        long triggerMillis = System.currentTimeMillis();
+        int streamReadTimeout = sharedPref.getInt("stream_read_timeout", 10) * 10000;
+
+        setAlarmVolume(minVolume);
+
+        Handler handler = new Handler();
+        Runnable runnable = new Runnable() {
+            boolean isPlaying = false;
+            long playStatedMillis = 0;
+            float currentVolume = minVolume;
+            @Override
+            public void run() {
+                if(BuildConfig.DEBUG) { Log.d(TAG, "Increase volume loop"); }
+
+                long totalElapsedMillis = System.currentTimeMillis() - triggerMillis;
+
+                try {
+                    isPlaying = itsPlayerService.isPlaying();
+                } catch (Exception e) {
+                    if (BuildConfig.DEBUG) { Log.d(TAG, "Couldn't get isPlaying"); }
+                }
+
+                if (isPlaying) {
+                    if (playStatedMillis == 0) {
+                        playStatedMillis = System.currentTimeMillis();
+                        setAlarmVolume(minVolume);
+                    }
+
+                    long elapsedMillis = System.currentTimeMillis() - playStatedMillis;
+                    float slowWakeProgress = (float) elapsedMillis / slowWakeMillis;
+
+                    if (currentVolume < maxVolume) {
+                        float newVolume = minVolume + Math.min(maxVolume, slowWakeProgress * volumeRange);
```

Comment:
The correct formula will be 
`float newVolume = Math.min(minVolume + slowWakeProgress * volumeRange, maxVolume);`
because with the current one you will get `minVolume + maxVolume` at peak.